### PR TITLE
Feature/add generic repository

### PR DESCRIPTION
**Pull request by Blackolf931**

### CODE REVIEW

File: TSP/BLL/AutoMapper/BllProfile.cs [hunk 1]
```diff
@@ -12,6 +12,8 @@ public BllProfile()
             CreateMap<Employee, EmployeeEntity>();
             CreateMap<OfficeEntity, Office>();
             CreateMap<Office, OfficeEntity>();
+            CreateMap<object, OfficeEntity>();
```
Comment (nantonov) [inline TSP/BLL/AutoMapper/BllProfile.cs:15]: It's wrong

File: TSP/BLL/Services/EmployeeService.cs [hunk 2]
```diff
@@ -29,22 +26,32 @@ public async Task<Employee> AddAsync(Employee employee)
...
             return _mapper.Map<IEnumerable<Employee>>(employee);
         }
 
         public async Task<Employee> GetEmployeeByIdAsync(int id)
         {
-            var employee = await _repository.GetByIdAsync(id);
+            var employee = await _repository.FindByIdAsync(id);
             var mappedEmployee = _mapper.Map<Employee>(employee);
-            if(employee is null)
+            if (employee is null)
```
Comment (nantonov) [inline TSP/BLL/Services/EmployeeService.cs:54]: Why this check is here?

File: TSP/BLL/Services/IEmployeeService.cs [hunk 3]
```diff
@@ -1,15 +1,16 @@
-﻿using BLL.Models;
+﻿
```
Comment (nantonov) [inline TSP/BLL/Services/IEmployeeService.cs:1]: For what this empty line?

File: TSP/DAL/Repositories/RepositoryBase.cs [hunk 4]
```diff
@@ -1,28 +1,47 @@
...
+            _repositoryContext = repositoryContext;
+        }
+        public async Task<IEnumerable<T>> FindAllAsync()
+        {
+            return await _repositoryContext.Set<T>().ToListAsync();
+        }
+
+        public async Task<T> FindByIdAsync(int id)
+        {
+            return await _repositoryContext.Set<Task<T>>().Find(id);
```
Comment (nantonov) [inline TSP/DAL/Repositories/RepositoryBase.cs:24]: Task<T> is not an entity

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/BLL/AutoMapper/BllProfile.cs (hunk 1 above)
Comment (Blackolf931) [reply TSP/BLL/AutoMapper/BllProfile.cs]: fix

File: TSP/BLL/Services/EmployeeService.cs (hunk 2 above)
Comment (Blackolf931) [reply TSP/BLL/Services/EmployeeService.cs]: fix

File: TSP/BLL/Services/IEmployeeService.cs (hunk 3 above)
Comment (Blackolf931) [reply TSP/BLL/Services/IEmployeeService.cs]: fix

File: TSP/DAL/Repositories/RepositoryBase.cs (hunk 4 above)
Comment (Blackolf931) [reply TSP/DAL/Repositories/RepositoryBase.cs]: fix

File: TSP/BLL/Services/EmpoloyeeServiceGetAdditionalInfo.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
+﻿using AutoMapper;
+using BLL.Models;
+using DAL.Interfaces;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace BLL.Services
+{
+    public class EmployeeServiceGettAdditionalInfo<T, TEntity> : GenericService<T,TEntity>, IGenericService<T,TEntity>
```
Comment (nantonov) [inline TSP/BLL/Services/EmpoloyeeServiceGetAdditionalInfo.cs:9]: Generic is not needed here

Comment (nantonov) [reply TSP/BLL/Services/EmpoloyeeServiceGetAdditionalInfo.cs:9]: Incorrect naming

File: TSP/TSP.API/Controllers/EmployeeController.cs [hunk 6]
```diff
@@ -9,16 +9,16 @@
 
 namespace TSP.API.Controllers
 {
-    public class EmployeeController : ControllerBase
+   /* public class EmployeeController : ControllerBase
```
Comment (nantonov) [inline TSP/TSP.API/Controllers/EmployeeController.cs:12]: This commented code mustn't reach a PR.

File: TSP/TSP.API/Controllers/OfficeController.cs [hunk 7]
```diff
@@ -13,55 +13,54 @@ namespace TSP.API.Controllers
     [Route("[controller]")]
     public class OfficeController : ControllerBase
     {
-        private readonly IOfficeService _service;
-        private readonly IMapper _mapper;
-        private readonly IValidator<OfficeAddViewModel> _validator;
-        private readonly IValidator<OfficeViewModel> _validatorViewModel;
+        /*  private readonly IMapper _mapper;
```
Comment (nantonov) [inline TSP/TSP.API/Controllers/OfficeController.cs:16]: Mustn't be here

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/TSP.API/Controllers/EmployeeController.cs (hunk 6 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/EmployeeController.cs:12]: if i uncomment there will be a build error

Comment (nantonov) [reply TSP/TSP.API/Controllers/EmployeeController.cs:12]: Fix the error

File: TSP/TSP.API/Controllers/OfficeController.cs (hunk 7 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/OfficeController.cs:16]: I'm trying to fix it

File: TSP/TSP.API/Controllers/EmployeeController.cs (hunk 6 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/EmployeeController.cs:12]: I'm trying to fix it

File: TSP/BLL/Services/EmpoloyeeServiceGetAdditionalInfo.cs (hunk 5 above)
Comment (Blackolf931) [reply TSP/BLL/Services/EmpoloyeeServiceGetAdditionalInfo.cs]: fix

File: TSP/BLL/Services/ServiceForEmployeeSetStrategy.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
+﻿using AutoMapper;
+using BLL.Models;
+using DAL.Entities;
+using DAL.Interfaces;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace BLL.Services
+{
+    public class ServiceForEmployeeSetStrategy : GenericService<Employee,EmployeeEntity>
```
Comment (nantonov) [inline TSP/BLL/Services/ServiceForEmployeeSetStrategy.cs:10]: EmployeeService

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/BLL/Services/EmpoloyeeServiceGetAdditionalInfo.cs (hunk 5 above)
Comment (Blackolf931) [reply TSP/BLL/Services/EmpoloyeeServiceGetAdditionalInfo.cs]: fix

File: TSP/TSP.API/Controllers/EmployeeController.cs (hunk 6 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/EmployeeController.cs]: fix

File: TSP/TSP.API/Controllers/OfficeController.cs (hunk 7 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/OfficeController.cs]: fix

File: TSP/BLL/Services/ServiceForEmployeeSetStrategy.cs (hunk 8 above)
Comment (Blackolf931) [reply TSP/BLL/Services/ServiceForEmployeeSetStrategy.cs]: fix

File: TSP/BLL/DI/DependencyRegistrar.cs [hunk 9]
```diff
@@ -12,13 +12,13 @@ public static class DependencyRegistrar
     {
         public static void RegistarBuisnessComponents(this IServiceCollection services, IConfiguration configuration)
         {
-            services.AddScoped<IEmployeeService, EmployeeService>();
-            services.AddScoped<IOfficeService, OfficeService>();
-            services.AddScoped<IEmployeeRepository, EmployeeRepository>();
-            services.AddScoped<IOfficeRepository, OfficeRepository>();
+            services.AddScoped(typeof(IRepositoryBase<>), typeof(RepositoryBase<>));
+            services.AddScoped(typeof(IGenericService<,>), typeof(GenericService<,>));
```
Comment (nantonov) [inline TSP/BLL/DI/DependencyRegistrar.cs:16]: Does it really needed?

File: TSP/TSP.API.XUnitTests/EmployeeServiceTests.cs [hunk 10]
```diff
@@ -15,24 +15,25 @@ namespace TSP.API.XUnitTests
 
     public class EmployeeServiceTests
     {
-        private readonly EmployeeService _sut;
-        private readonly Mock<IEmployeeRepository> _employeeRepoMock = new();
+        private GenericService<Employee, EmployeeEntity> _sut;
```
Comment (nantonov) [inline TSP/TSP.API.XUnitTests/EmployeeServiceTests.cs:18]: Why generic?

File: TSP/TSP.API/Startup.cs [hunk 11]
```diff
@@ -1,5 +1,6 @@
 using BLL.AutoMapper;
 using BLL.DI;
+using BLL.Services;
```
Comment (nantonov) [inline TSP/TSP.API/Startup.cs:3]: For what?

File: TSP/TSP.API/Validators/EmployeeViewValidator.cs [hunk 12]
```diff
@@ -11,7 +7,6 @@ public class EmployeeViewValidator : AbstractValidator<EmployeeViewModel>
     {
         public EmployeeViewValidator()
```
Comment (nantonov) [inline TSP/TSP.API/Validators/EmployeeViewValidator.cs:8]: When this validator will be called?

File: TSP/TSP.API/ViewModels/EmployeeViewModel.cs [hunk 13]
```diff
@@ -1,14 +1,13 @@
 ﻿namespace TSP.API.ViewModels
 {
     public class EmployeeViewModel
-    {
-        public int Id { get; set; }
-        public string Name { get; set; }
-        public string SecondName { get; set; }
-        public string Patronomic { get; set; }
+    { 
+        public string? Name { get; set; }
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/EmployeeViewModel.cs:5]: Is name can be null?

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/TSP.API/Validators/EmployeeViewValidator.cs (hunk 12 above)
Comment (Blackolf931) [reply TSP/TSP.API/Validators/EmployeeViewValidator.cs]: deleted

File: TSP/TSP.API/ViewModels/EmployeeViewModel.cs (hunk 13 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/EmployeeViewModel.cs]: fix

File: TSP/TSP.API/Startup.cs (hunk 11 above)
Comment (Blackolf931) [reply TSP/TSP.API/Startup.cs]: deleted

File: TSP/TSP.API.XUnitTests/EmployeeServiceTests.cs (hunk 10 above)
Comment (Blackolf931) [reply TSP/TSP.API.XUnitTests/EmployeeServiceTests.cs]: fix

File: TSP/BLL/DI/DependencyRegistrar.cs (hunk 9 above)
Comment (Blackolf931) [reply TSP/BLL/DI/DependencyRegistrar.cs]: deleted

File: TSP/BLL/DI/DependencyRegistrar.cs [hunk 14]
```diff
@@ -12,17 +12,16 @@ public static class DependencyRegistrar
     {
         public static void RegistarBuisnessComponents(this IServiceCollection services, IConfiguration configuration)
         {
-            services.AddScoped<IEmployeeService, EmployeeService>();
-            services.AddScoped<IOfficeService, OfficeService>();
-            services.AddScoped<IEmployeeRepository, EmployeeRepository>();
-            services.AddScoped<IOfficeRepository, OfficeRepository>();
+            services.AddScoped(typeof(IRepositoryBase<>), typeof(RepositoryBase<>));
```
Comment (nantonov) [inline TSP/BLL/DI/DependencyRegistrar.cs:15]: It's not a business component

File: TSP/BLL/Services/EmployeeService.cs [hunk 15]
```diff
@@ -1,70 +1,45 @@
...
-        public EmployeeService(IEmployeeRepository repository, IMapper mapper, IEnumerable<IStrategy> strategy)
+        public EmployeeService(IRepositoryBase<EmployeeEntity> repository, IMapper mapper, IEnumerable<IStrategy> strategy) : base(repository, mapper)
         {
             _repository = repository;
             _mapper = mapper;
             _strategy = strategy;
         }
 
-        public async Task<Employee> AddAsync(Employee employee)
+        public override async Task<Employee> GetByIdAsync(int id)
```
Comment (nantonov) [inline TSP/BLL/Services/EmployeeService.cs:23]: Simplify the method

File: TSP/BLL/Services/GenericService.cs [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        public async Task<bool> DeleteByIdAsync(int id)
+        {
+            var entityObject = await _repository.FindByIdAsync(id);
+            if (entityObject is null)
+            {
+                return false;
+            }
+            else
+            {
+                var mapped = _mapper.Map<TEntity>(entityObject);
```
Comment (nantonov) [inline TSP/BLL/Services/GenericService.cs:37]: Are you mapping entity to entity?

File: TSP/BLL/Services/IEmployeeService.cs [hunk 17]
```diff
@@ -1,15 +1,15 @@
 ﻿using BLL.Models;
+using DAL.Entities;
+using System;
 using System.Collections.Generic;
+using System.Linq;
```
Comment (nantonov) [inline TSP/BLL/Services/IEmployeeService.cs:5]: Too much usings

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/BLL/DI/DependencyRegistrar.cs (hunk 14 above)
Comment (Blackolf931) [reply TSP/BLL/DI/DependencyRegistrar.cs]: fix

File: TSP/BLL/Services/EmployeeService.cs (hunk 15 above)
Comment (Blackolf931) [reply TSP/BLL/Services/EmployeeService.cs]: fix

File: TSP/BLL/Services/GenericService.cs (hunk 16 above)
Comment (Blackolf931) [reply TSP/BLL/Services/GenericService.cs]: fix

File: TSP/BLL/Services/IEmployeeService.cs (hunk 17 above)
Comment (Blackolf931) [reply TSP/BLL/Services/IEmployeeService.cs]: fix

File: TSP/BLL/Services/EmployeeService.cs [hunk 18]
```diff
@@ -1,70 +1,42 @@
...
 using System.Threading.Tasks;
 
 namespace BLL.Services
 {
-    public class EmployeeService : IEmployeeService
+    public class EmployeeService : GenericService<Employee?, EmployeeEntity>, IEmployeeService
     {
-        private readonly IEmployeeRepository _repository;
-        private readonly IMapper _mapper;
+        public IRepositoryBase<EmployeeEntity> _repository;
```
Comment (nantonov) [inline TSP/BLL/Services/EmployeeService.cs:12]: Why public?

File: TSP/BLL/Services/OfficeService.cs [hunk 19]
```diff
@@ -1,51 +1,19 @@
...
-using System.Threading.Tasks;
 
 namespace BLL.Services
 {
-    public class OfficeService : IOfficeService
+    public class OfficeService : GenericService<Office, OfficeEntity>, IOfficeService
     {
-        private readonly IOfficeRepository _repository;
-        private readonly IMapper _mapper;
+        public IRepositoryBase<OfficeEntity> _repository;
```
Comment (nantonov) [inline TSP/BLL/Services/OfficeService.cs:10]: Why public?

File: TSP/DAL/DAL.csproj [hunk 20]
```diff
@@ -1,15 +1,17 @@
...
   <ItemGroup>
 
     <PackageReference Include="EntityFramework" Version="6.4.4" />
-    <PackageReference Include="AutoMapper" Version="10.1.1" />
     <PackageReference Include="EntityFramework" Version="6.4.4" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.9" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.10" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="5.0.10" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="5.0.0" />
     <PackageReference Include="Ninject" Version="3.3.4" />
```
Comment (nantonov) [inline TSP/DAL/DAL.csproj:15]: Are you using it?

File: TSP/TSP.API/Controllers/EmployeeController.cs [hunk 21]
```diff
@@ -32,7 +32,7 @@ public async Task<ActionResult<IEnumerable<EmployeeViewModel>>> GetAllEmployeeAs
         [HttpGet("GetEmployeeById")]
         public async Task<ActionResult<Employee>> GetEmployeeByIdAsync([FromQuery] int id)
```
Comment (nantonov) [inline TSP/TSP.API/Controllers/EmployeeController.cs:33]: EmployeeViewModel

File: TSP/TSP.API/Controllers/OfficeController.cs [hunk 22]
```diff
@@ -13,55 +13,53 @@ namespace TSP.API.Controllers
...
 
-        public OfficeController(IOfficeService service, IMapper mapper, IValidator<OfficeAddViewModel> validator, IValidator<OfficeViewModel> validatorViewModel)
-        {
-            _service = service;
-            _mapper = mapper;
-            _validator = validator;
-            _validatorViewModel = validatorViewModel;
-        }
+          [HttpGet("GetOfficeById")]
+          public async Task<ActionResult<Office>> GetOfficeById(int id)
```
Comment (nantonov) [inline TSP/TSP.API/Controllers/OfficeController.cs:37]: ViewModel

File: TSP/TSP.API/Controllers/OfficeController.cs [hunk 23]
```diff
@@ -13,55 +13,53 @@ namespace TSP.API.Controllers
...
-            return Ok(mapped);
-        }
+          [HttpDelete("DeleteById")]
+          public async Task<ActionResult<bool>> DeleteByIdAsync(int id)
+          {
+              await _service.DeleteByIdAsync(id);
+              return Ok();
+          }
+          [HttpPost("AddOffice")]
+          public async Task<ActionResult<Office>> AddOffice([FromBody] OfficeAddViewModel model)
```
Comment (nantonov) [inline TSP/TSP.API/Controllers/OfficeController.cs:49]: ViewModel

File: TSP/TSP.API/Controllers/OfficeController.cs [hunk 24]
```diff
@@ -13,55 +13,53 @@ namespace TSP.API.Controllers
...
-        [HttpPut("UpdateOffice")]
-        public async Task<ActionResult<Office>> UpdateOffice([FromQuery] int id, [FromBody] OfficeViewModel model)
-        {
-            await _validatorViewModel.ValidateAndThrowAsync(model);
-            var mapped = _mapper.Map<Office>(model);
-            mapped.Id = id;
-            return Ok(await _service.UpdateOfficeByAsync(mapped));
-        }
+          [HttpPut("UpdateOffice")]
+          public async Task<ActionResult<Office>> UpdateOffice([FromQuery] int id, [FromBody] OfficeViewModel model)
```
Comment (nantonov) [inline TSP/TSP.API/Controllers/OfficeController.cs:57]: ViewModel

File: TSP/TSP.API/ViewModels/EmployeeViewModel.cs [hunk 25]
```diff
@@ -1,8 +1,8 @@
 ﻿namespace TSP.API.ViewModels
 {
+#nullable disable
     public class EmployeeViewModel
     {
-        public int Id { get; set; }
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/EmployeeViewModel.cs:5]: Id should be presented

File: TSP/TSP.API/ViewModels/EmployeeViewModel.cs [hunk 26]
```diff
@@ -1,8 +1,8 @@
 ﻿namespace TSP.API.ViewModels
 {
+#nullable disable
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/EmployeeViewModel.cs:3]: After disable - enable

File: TSP/TSP.API/ViewModels/OfficeAddViewModel.cs [hunk 27]
```diff
@@ -1,16 +1,10 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading.Tasks;
-
-namespace TSP.API.ViewModels
+﻿namespace TSP.API.ViewModels
 {
+#nullable disable
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/OfficeAddViewModel.cs:3]: Enable is missed

File: TSP/TSP.API/ViewModels/OfficeViewModel.cs [hunk 28]
```diff
@@ -1,16 +1,10 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading.Tasks;
-
-namespace TSP.API.ViewModels
+﻿namespace TSP.API.ViewModels
 {
+#nullable disable
```
Comment (nantonov) [inline TSP/TSP.API/ViewModels/OfficeViewModel.cs:3]: Enable

File: TSP/DAL/DAL.csproj [hunk 29]
```diff
@@ -1,15 +1,17 @@
...
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <TargetFramework>net5.0</TargetFramework>
+    <Nullable>enable</Nullable>
   </PropertyGroup>
 
   <ItemGroup>
 
     <PackageReference Include="EntityFramework" Version="6.4.4" />
```
Comment (nantonov) [inline TSP/DAL/DAL.csproj:10]: Is it needed here?

File: TSP/TSP.API/TSP.API.csproj [hunk 30]
```diff
@@ -2,18 +2,19 @@
 
   <PropertyGroup>
     <TargetFramework>net5.0</TargetFramework>
+    <Nullable>enable</Nullable>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="AutoMapper" Version="10.1.1" />
     <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="8.1.1" />
     <PackageReference Include="EntityFramework" Version="6.4.4" />
```
Comment (nantonov) [inline TSP/TSP.API/TSP.API.csproj:11]: Is it needed here?

**Review by nantonov — CHANGES_REQUESTED** [review]

File: TSP/BLL/Services/EmployeeService.cs (hunk 18 above)
Comment (Blackolf931) [reply TSP/BLL/Services/EmployeeService.cs]: fix

File: TSP/BLL/Services/OfficeService.cs (hunk 19 above)
Comment (Blackolf931) [reply TSP/BLL/Services/OfficeService.cs]: fix

File: TSP/TSP.API/Controllers/EmployeeController.cs (hunk 21 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/EmployeeController.cs]: fix

File: TSP/TSP.API/Controllers/OfficeController.cs (hunk 22 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/OfficeController.cs]: fix

File: TSP/TSP.API/Controllers/OfficeController.cs (hunk 23 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/OfficeController.cs]: fix

File: TSP/TSP.API/Controllers/OfficeController.cs (hunk 24 above)
Comment (Blackolf931) [reply TSP/TSP.API/Controllers/OfficeController.cs]: fix

File: TSP/TSP.API/ViewModels/EmployeeViewModel.cs (hunk 25 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/EmployeeViewModel.cs:5]: fix

File: TSP/TSP.API/ViewModels/OfficeAddViewModel.cs (hunk 27 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/OfficeAddViewModel.cs:3]: fix

File: TSP/TSP.API/ViewModels/OfficeViewModel.cs (hunk 28 above)
Comment (Blackolf931) [reply TSP/TSP.API/ViewModels/OfficeViewModel.cs:3]: fix

File: TSP/DAL/DAL.csproj (hunk 29 above)
Comment (Blackolf931) [reply TSP/DAL/DAL.csproj:10]: I only found this way to enable this. What should i do to solve this problem?

File: TSP/DAL/DAL.csproj (hunk 20 above)
Comment (Blackolf931) [reply TSP/DAL/DAL.csproj]: fix

File: TSP/DAL/DAL.csproj (hunk 29 above)
Comment (Blackolf931) [reply TSP/DAL/DAL.csproj]: fix

File: TSP/TSP.API/TSP.API.csproj (hunk 30 above)
Comment (Blackolf931) [reply TSP/TSP.API/TSP.API.csproj]: fix

**Review by nantonov — APPROVED** [review]

**Pull request merged**